src/libstd/path.rs

```diff
@@ -639,6 +639,18 @@ pub struct Iter<'a> {
     inner: Components<'a>,
 }
 
+#[stable(feature = "path_components_debug", since = "")]
+impl<'a> fmt::Debug for Components<'a> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let components = self.as_path()
+                             .components()
+                             .collect::<Vec<_>>();
```

Comment:
Just make a helper struct that implements debug and outputs `[...]`. That is,

```rust
f.debug_tuple("Components").field(&DebugHelper(self)).finish();
```

where `DebugHelper` implements `Debug`.